Add search support to dropdown menus

Mati365 opened the pull request:
Feature (ui): Add search support to dropdown menus.

Part of https://github.com/ckeditor/ckeditor5/issues/16311.

---

### Related PRs

Focus change PR: https://github.com/ckeditor/ckeditor5/pull/16642
Base dropdown component PR: https://github.com/ckeditor/ckeditor5/pull/16631
AI integration PR: https://github.com/cksource/ckeditor5-commercial/pull/6213

### Screens

![Menu](https://github.com/ckeditor/ckeditor5/assets/3949262/cd9f319b-d00b-41c2-896c-0de58c5c52ac)
![Search](https://github.com/ckeditor/ckeditor5/assets/3949262/d67b6f19-210a-4380-bc11-724cb19ca640)

### Further context

This PR adds search input to whole dropdown menus implementation. It reduced the size of main PR.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/search/createtreefromflattendropdownmenuslist.ts:
```diff
@@ -0,0 +1,176 @@
+/**
+ * The maximum depth of a dropdown menu tree.
+ */
+type MaxDropdownTreeMenuDepth = 6;
```
It's used. Your comment is simply wrong. Bad rabbit, very bad rabbit.

Mati365 replied:
No more carrot this week.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/search/walkoverdropdownmenutreeitems.ts:
```diff
@@ -0,0 +1,168 @@
+ * @template K The type of dropdown menu node kind.
+ * @template Extend The type of additional data that can be passed to the walkers.
+ */
+export type DropdownMenuViewsTreeWalkerMetadata<
+	K extends DropdownMenusViewsTreeNodeKind = DropdownMenusViewsTreeNodeKind,
+	Extend = unknown
+> = {
+	visitor: DropdownMenuViewsTreeVisitor;
+	node: ExtractDropdownMenuViewTreeNodeByKind<K, Extend>;
+	parents: NonEmptyArray<DropdownMenusViewsTreeNode<Extend>>;
+	parent: DropdownMenusViewsTreeNode<Extend>;
+};
```
Rabbit, your suggestion is identical as code that is already present in repo

scofalik commented:
Thanks for providing an overview of changes introduced by the PR. I am in the middle of the review but I have some questions about what you wrote.

> ```javascript
> // Somewhere later in the code.
> view.appendTopLevelChild( {
>   menu: 'Menu 2',
>   children: [
>     new DropdownMenuListItemButtonView( locale, 'Item 2' )
>   ]
> } );
> ```

Do we need this for any specific functionality right now? The two features that for sure will use the new dropdowns are Merge fields and AI assistant. Both features have a static list of categories and items.

Much of what you wrote and explained is related to adding items dynamically. It wasn't in the scope of this component and maybe it can be simplified. One, because it's easier to review and close a PR when there's less code. Two, because we don't want to provide a public API too hastily. Three, because there's a high chance this code will be unified and/or refactored anyway in a few quarters. And four, because it might be beneficial in another area:

> 1.  `tree` / `menus` getter calls might be expensive for huge menus (> 500 entries). It's not real case and at this moment there are no actions required to be performed in this scope.

We have a real case for this, as we have reports from some users that they are looking to use as many as hundreds of merge fields. So, I am concerned about performance here.

Mati365 commented:
@scofalik 

> Do we need this for any specific functionality right now? The two features that for sure will use the new dropdowns are Merge fields and AI assistant. Both features have a static list of categories and items.

Nope. These methods simplify mocking of the data in the tests and basically are super small in terma of implementation (it's rather an allias for internal method). What do you think about marking them as internal?

> We have a real case for this, as we have reports from some users that they are looking to use as many as hundreds of merge fields. So, I am concerned about performance here

:o afaik menu is not designed to hold that amount od data, we have to discuss that

Mati365 commented:
@scofalik Lazy initialization, menu cache and found items limit added. It means that we can handle ~800 items without drastically reducing startup / navigation performance. 

There are few downsides:

1. The search engine needs to perform full initialization of all lazy menus during the first search (it's required to walk through all button or other "flat" non-menus entries).
2. I kept it behind the flag. Besides that, the performance is much better, it makes menus equally harder to unit test in other integrations because menu is not fully initialized. 
3. The most crucial one - lazy initialization must be disabled in submenus where their buttons `isEnabled` state is controlled by external observables. The primary example might be `AI Commands` view with `enabledCommandsIds`. Each command button might be dynamically disabled, and search engine detects that situations.

Mati365 commented:
Notes after discussion with @scofalik / @oleq. 

1. Idea of editor panel placement in `ck-body-wrapper` is ok. 
2. We split dropdown implementation into multiple phases:
2.1. Add scrolling / resizing support (it is already broken). It's **current** iteration.
2.2. Add better support for truncating dropdown menus when editor is not wide enough and has some overflow hidden styling.
2.3. Improve blur / focus support (it's probably related to every dropdown in editor).
2.4. Adding various demos to manual tests.

Mati365 commented:
I improved a bit DX of menu declaration definition. It's now possible to use definition parser used in root list view in submenus, which simplifies testing. `appendTopLevelChild` has been replaced with `factory.appendChild` call.

Mati365 commented:
I moved dump menu logic to test utils and renamed `kind` to `type` in dropdown tree.

Mati365 commented:
Follow-up ticket: https://github.com/ckeditor/ckeditor5/issues/16613

scofalik reviewed (changes requested):
I reviewed the PR. Despite committing quite some time to it, I'd say that was just only initial review. I haven't touched many of the included files.

I believe this PR is overwhelmingly big and it includes many changes/functionalities, which makes it impossible to fully grasp and analyze it. It's hard to build confidence with such PR. It's also overengineered in a few places.

One thing that bothers me is that the current menu bar implementation is a bit overcomplicated too (with the amount of classes it provides). Since this PR was supposed to base on the menu bar, it inherited that overcomplexity.

In order to ship some functionality in a reasonable time, I propose a fixing plan for this PR. I guess the only other option would be to start over clean.

First of all, let's remove all stuff related to search, search input, filtered view, related utils/mechanisms, and CSSes. Let's focus on introducing just the component for the nested dropdowns, similar what we have in the menu bar. Things that are directly related to the core functionality, like e.g. creating panels in body-container to allow max-height, they should stay. We will add search capabilities in next step.

Second, let's try to remove all the typings, guards, utils and "it will be useful"-things that are not necessarily needed. If something is used just in one class, move it to that class file and have it as a local function / private method. I think that _some_ of the _utilities_ can be helpful, like e.g. access to the flattened list of items in the whole dropdown structure. This can stay if it is still used by anything in the code (after other simplifications are introduced). But in general, let's keep these utils at minimum.

Third, following above, let's try to have less files total. I hope we will have less utils and they will all go to one file (`dropdown/menu/utils.ts`). Types or events declarations can go to the files they are strictly related to. If we really need to, we can also have `dropdown/menu/typings.ts`. Menu bar has 9 files total, I hope we can achieve similar quantity here.

Fourth, implement other simplifications I mentioned in comments to this review.

Fifth, I'll repeat again, let's keep the amount of changes to the minimum.

I wrote that I made only "initial review". I didn't review search and filter view related files at all. In the rest of the files, I mostly commented on things I didn't understand or I felt are overcomplicated or maybe not necessary, but I didn't focus enough to make sure that everything is bug-free etc. So there will have to be another review, more in-depth, hopefully the new PR will be 1/3rd of the size of this.

Some other tips:

"Old" dropdowns, which we have in toolbars, they allow to have everything in them. But in this case, we are implementing a solution which is very well defined and it does not need to be very open-ended. We just need lists with two things: buttons that perform actions, and buttons that open more lists. We should "live" only in `dropdown/menu` space. Of course, extending some base classes like `View` or `ListView` (if it helps) but we can be specific on what we allow in "items". Basically, when you iterate over items, you should always have only two options: "leaf" button, or "dropdown-trigger" button. MAYBE separator. Finally, "menu" dropdowns, do not need to replace current dropdowns, so let's focus on the two use cases we have, i.e. AI Assistant and Merge Fields.

When I think about AI Assistant and Merge Fields, how would I like to create the component for them, it's something that could be simplified to:

```js
// Config value, post-processed to fit `DropdownMenu` definition structure.
const definition = [
    {
        // Merge field category.
        label: 'Customer',
        id: 'customer',
        items: [
            // Button, upon clicking inserts a merge field.
            {
                label: 'Customer name',
                id: 'customer.name' // Merge field name.
            },
            {
                label: 'Customer address',
                id: 'customer.address'
            },
            // ... More merge fields
        ]
    },
    // ... More merge fields categories
    // Could have top-level merge fields too.
];

// When leaf button is clicked, `onExecute` is called with its id.
const onExecute = buttonId => editor.execute( 'insertMergeField', buttonId );

// Create dropdown menu structure. This could be called `DropdownMenuView` to keep the convention,
// or if we don't want to have such heavy view, it could be created by an utility function/factory.
const mergeFieldsDropdown = new DropdownMenu( editor.locale, definition, onExecute );

// Create toolbar button:
const button = new Button( ... );

// Show dropdown when button is clicked, simplified, of course, but that's outside of the scope of this PR.
button.on( 'execute', () => {
    mergeFieldsDropdown.isOpen = true;
} );
```

I would also try to get rid of the "intermediate" classes like `DropdownMenuListItemView` or `DropdownMenuPanelView`, cause it would seem to me that all we need is: button and list. MAYBE separator. I am not sure if that's 100% possible, but I'd try to do it.

scofalik commented:
FYI: I removed some of the comments that got to the review accidentally and I haven't seen them because the code changed (I started the review a long time ago...).

scofalik commented:
Just to make it clear: I wouldn't go fixing the comments one after another. E.g. some of the files/classes that I commented may not end up in the final PR after simplifications, or they may got refactored to the point that my comment is outdated. So, first make the plan what is to be kept, what should be refactored, and what should be removed. Then see if my comments still apply.

You can start over if that's easier for you and copy-paste parts of the code from this branch.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/definition/dropdownmenudefinitiontypings.ts:
```diff
@@ -0,0 +1,34 @@
+ * Represents a child definition of a dropdown menu.
+ * It can be either a `DropdownMenuDefinition` or a `DropdownMenuOrFlatItemView`.
+ */
+export type DropdownMenuChildDefinition = DropdownMenuDefinition | DropdownMenuOrFlatItemView;
```
> I don't like the order of things in this file.

Reordered, definition object is reduced.

> Do we have an actual requirement where we must pass instances

At least two reasons:

1. In some of our plugins, like in [AI](https://github.com/cksource/ckeditor5-commercial/blob/master/packages/ckeditor5-ai/src/aiassistantui.ts#L576) there is filterable view. From what I can see, it used to operate on plain instances that used to be a bit customized (at least manually created). They usually inherit from `ListItemView` to store some additional information about commands that are later used to control `isOn` state right [there](https://github.com/cksource/ckeditor5-commercial/blob/master/packages/ckeditor5-ai/src/ui/dropdown/aicommandslistview.ts#L47). The other thing is that it also watches on `execute` on such items like [here](https://github.com/cksource/ckeditor5-commercial/blob/master/packages/ckeditor5-ai/src/aiassistantui.ts#L255C24-L255C31). 
2. It simplifies binding `isOn` / `isEnabled` states in filter view. Let's assume we are filtering for `Butter`, it displays found menu items that `isEnabled` state is bound to these ancestor menu item instances. It would be not possible in JSON approach, where the listener should be implemented for both filtering view and menu view.

Mati365 replied:
After discussion, we decided I'll keep possibility to `View` to dropdown menu and add additional flat item definition entry with `onExecute`.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/utils/dropdownmenubehaviors.ts:
```diff
@@ -0,0 +1,189 @@
+	menuView.panelView.element!.contains( document.activeElement ) ||
+		menuView.element!.contains( document.activeElement );
+
+export const DropdownRootMenuBehaviors = {
```
@scofalik Unfortunately it makes no difference in terms of manual testing. Dropdowns are different environment and I had to manually adjust these functions so they are not 100% copied.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/definition/dropdownmenudefinitiontypings.ts:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Represents the children of a dropdown menu.
+ */
+export type DropdownMenuChildrenDefinition = Array<DropdownMenuChildDefinition>;
```
@scofalik I removed many types from this file. Now it has bare minimum and looks better. Please reopen if there is still too many types.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/typings.ts:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Represents a view for a non-focusable flat item in a dropdown menu.
+ */
+export type DropdownMenuNonFocusableFlatItemView = DropdownMenuListItemButtonView;
```
@scofalik  I'm resolving, please reopen if it's not enough.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/utils/dropdownmenupositioningfunctions.ts:
Yep, it's probably a bigger topic and needs more attention. At this moment, I resolve this comment.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/utils.ts:
```diff
@@ -106,13 +106,16 @@ import ListItemGroupView from '../list/listitemgroupview.js';
 	ButtonClassOrInstance:
-		( new ( locale?: Locale ) => DropdownButton & FocusableView ) | DropdownButton & FocusableView = DropdownButtonView
+		( new ( locale?: Locale ) => DropdownButton & FocusableView ) | DropdownButton & FocusableView = DropdownButtonView,
+	defaultBehaviorAttributes: DefaultBehaviorAttributes = {}
```
Renamed

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/dropdownmenuview.ts:
```diff
@@ -0,0 +1,380 @@
+		const { panelView } = this;
+		const { body } = this.editor.ui.view;
+
+		// Let the menu control the position of the panel. The position must be updated every time the menu is open.
```
Removed

Mati365 commented:
@scofalik 

## Size of PR

> I believe this PR is overwhelmingly big and it includes many changes/functionalities, which makes it impossible to fully grasp and analyze it. It's hard to build confidence with such PR. It's also overengineered in a few places.

Yes, it's because it's menu with some additional features like search integration and scrolling support, which are not present in menubar alone, and added a bit of complexity. Keep in mind that the current implementation of menubar is approximately 7k lines long, without search functionality but with some other enhancements, so the current solution proposed in this PR _is_ compact even if it has a lot of examples in the comments.  

## Remove search function

> First of all, let's remove all stuff related to search, search input, filtered view, related utils/mechanisms, and CSSes. Let's focus on introducing just the component for the nested dropdowns, similar what we have in the menu bar. Things that are directly related to the core functionality, like e.g. creating panels in body-container to allow max-height, they should stay. We will add search capabilities in next step.

There are three problems with that:

#### Dropdown menus is designed around search

I designed a whole solution to keep implementation of search as small as I can, so basically the whole search feature is present in `filterview` directory which contains not so advanced components at all. I'm talking about these:

1. https://github.com/ckeditor/ckeditor5/pull/16314/files#diff-df3750388978cd16ceb1c31a216bfcf5b90d6a000ee5b4506767ac219fd5ec1fR1
2. https://github.com/ckeditor/ckeditor5/pull/16314/files#diff-41402c4b4c3d967f87213503412bb1476326838b31908c4cd7c12a1b0aa8ddca

which uses these helpers:

1. https://github.com/ckeditor/ckeditor5/pull/16314/files#diff-6fc38ee15fce051137018358bc00b15d65e7387685ee74d4f5bd8f3da73dd640R11
2. https://github.com/ckeditor/ckeditor5/pull/16314/files#diff-6fc38ee15fce051137018358bc00b15d65e7387685ee74d4f5bd8f3da73dd640R131

So around 20% of this PR will be reduced and of course, we can remove them, but I'm quite sure that they will remain in identical form finally.  To summarize: removal of search will not result in removal `tree` idea and any other things. 

#### There are pending changes in merge tags

There are few conflict places that modified the behavior of dropdowns in similar way the dropdown menus did. Like this one:

1. Highlight button view's location has been moved. Maybe it's behavior has been modified. https://github.com/cksource/ckeditor5-commercial/pull/6345/files#diff-70fcbed8dd0be04a9fc4c6a4179fb52eb188bb533dff9c20bde37f7cdd8d3f3e
2. Filtering has been changes https://github.com/cksource/ckeditor5-commercial/pull/6345/files#diff-9e2c8b176014c2617cc2850f313d0e75ca0167c6d9be19a37eaeb5ccfee715b6R67

I believe that there will be merge conflicts, and we have to resolve it as fast as we can. Splitting PR into two will delay it.

> Third, following above, let's try to have less files total. I hope we will have less utils and they will all go to one file (dropdown/menu/utils.ts). Types or events declarations can go to the files they are strictly related to. If we really need to, we can also have dropdown/menu/typings.ts. Menu bar has 9 files total, I hope we can achieve similar quantity here.

I'd suggest not doing that. While I get argument about "file invasion" I also see that testing and debugging huge "util file of truth" is even worse. I added a lot of examples and docs explaining what these helpers are doing, and I think that "util file of truth" will be too huge to maintain. 

At this moment I removed not necessary guards and typings and moved many of them to destination files. 

#### There is pending PR that unifies `isOn` state and focus

I'm talking about this one: https://github.com/ckeditor/ckeditor5/pull/16573. This PR does not include any change from such PR and needs to be unified. Making focus work is not a trivial thing according to our scrollable dropdown menus behavior, and it'll take time. 

## Change in DSL format

> When I think about AI Assistant and Merge Fields, how would I like to create the component for them, it's something that could be simplified to:

```tsx
// Config value, post-processed to fit `DropdownMenu` definition structure.
const definition = [
    {
        // Merge field category.
        label: 'Customer',
        id: 'customer',
        items: [
            // Button, upon clicking inserts a merge field.
            {
                label: 'Customer name',
                id: 'customer.name' // Merge field name.
            },
            {
                label: 'Customer address',
                id: 'customer.address'
            },
            // ... More merge fields
        ]
    },
    // ... More merge fields categories
    // Could have top-level merge fields too.
];

// When leaf button is clicked, `onExecute` is called with its id.
const onExecute = buttonId => editor.execute( 'insertMergeField', buttonId );

// Create dropdown menu structure. This could be called `DropdownMenuView` to keep the convention,
// or if we don't want to have such heavy view, it could be created by an utility function/factory.
const mergeFieldsDropdown = new DropdownMenu( editor.locale, definition, onExecute );

// Create toolbar button:
const button = new Button( ... );

// Show dropdown when button is clicked, simplified, of course, but that's outside of the scope of this PR.
button.on( 'execute', () => {
    mergeFieldsDropdown.isOpen = true;
} );
```

This format is not ideal, and it's blocker for customization of menu bar items. I described it briefly it comments above. Long story short: 

1. Some menu items are customized, especially the file dialog button. It inherits `ButtonView` and modifies it somehow. We had to add mixin that allow inheritance of such button from custom base class named `ListButtonView`. That inheritance above basically kills this approach. 
2. Some commands store additional metadata to perform `execute` command using inheritance from `ListItemView`. Such like AI commands. I've seen that in AI it's probably present in other places that use search-first approach.

I'd suggest to keep possibility to pass instance (for such plugins like AI where it works just fine), and add simple JSON definition for buttons with `onExecute` handler. 

It'll fix both problems and not increase complexity.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/definition/dropdownmenulistdefinitionfactory.ts:
```diff
@@ -0,0 +1,258 @@
+	 * @param parentMenuView The parent menu view.
+	 * @returns The registered menu or reused instance.
+	 */
+	private _createFocusableDefinitionChild(
```
I renamed this function, please reopen if it's still wrong.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/dropdownmenurootlistview.ts:
```diff
@@ -0,0 +1,258 @@
+ * const view = new DropdownMenuRootListView( editor, [
+ * 	{
+ * 		menu: 'Menu 1',
+ * 		children: [
+ * 			new DropdownMenuListItemButtonView( locale, 'Item' ),
+ * 			new ListSeparatorView( locale ),
+ * 			{
+ * 				menu: 'Menu 1.1',
+ * 				children: [
+ * 					new DropdownMenuListItemButtonView( locale, 'Nested Item' ),
+ * 				]
+ * 			}
+ * 		]
```
@scofalik This approach, like that one described in my comment, has been used across the project for a while. Basically, 50% of the menu integrations _do_ access `MenuListItem` constructors. Generally speaking, it's nothing new from what I've seen. In other comments, I described why instances are allowed here, though.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/dropdownmenuview.ts:
```diff
@@ -0,0 +1,380 @@
+		const { panelView, buttonView } = this;
+
+		// Let the menu control the position of the panel. The position must be updated every time the menu is open.
+		this.on<ObservableChangeEvent<boolean>>( 'change:isOpen', ( evt, name, isOpen ) => {
```
You are correct. It should be placed together.

Mati365 commented on packages/ckeditor5-core/src/typings.ts:
```diff
@@ -14,3 +14,19 @@ export type NonEmptyArray<A> = Array<A> & {
+			T[K] extends string ? Readonly<T[K]>
+				: T[K] extends Array<infer A> ? Readonly<Array<DeepReadonly<A, Increment<Level>>>>
+					: DeepReadonly<T[K], Increment<Level>>;
+	}>;
```
Removed after discussion.

Mati365 commented on packages/ckeditor5-ui/src/dropdown/menu/dropdownmenulistitemview.ts:
```diff
@@ -0,0 +1,54 @@
+	 */
+	public readonly flatItemOrNestedMenuView: DropdownMenuView | DropdownMenuFocusableFlatItemView;
+
+	constructor(
```
Renamed to `childView`.
